File: src/test/java/fr/adrienbrault/idea/symfony2plugin/tests/mcp/GenerateSymfonyServiceDefinitionTest.java
package fr.adrienbrault.idea.symfony2plugin.tests.mcp;

import fr.adrienbrault.idea.symfony2plugin.mcp.service.ServiceDefinitionGenerator;
import fr.adrienbrault.idea.symfony2plugin.tests.SymfonyLightCodeInsightFixtureTestCase;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class GenerateSymfonyServiceDefinitionTest extends SymfonyLightCodeInsightFixtureTestCase {

    @Override
    public void setUp() throws Exception {
        super.setUp();
        myFixture.configureFromExistingVirtualFile(myFixture.copyFileToProject("classes.php"));
    }

    protected String getTestDataPath() {
        return "src/test/java/fr/adrienbrault/idea/symfony2plugin/tests/action/ui/fixtures";
    }

    public void testGenerateYamlDefault() {
        // generate_symfony_service_definition('Foo\\Bar')
        ServiceDefinitionGenerator generator = new ServiceDefinitionGenerator(getProject());
        String result = generator.generate("Foo\\Bar", ServiceDefinitionGenerator.OutputType.YAML);

        assertNotNull(result);
        assertTrue(result.contains("Foo\\Bar:"));
    }

    public void testGenerateYamlWithClassNameAsIdTrue() {
        // generate_symfony_service_definition('Foo\\Bar', 'yaml', true)
        ServiceDefinitionGenerator generator = new ServiceDefinitionGenerator(getProject());
        String result = generator.generate("Foo\\Bar", ServiceDefinitionGenerator.OutputType.YAML, true);

        assertNotNull(result);
        assertTrue(result.contains("Foo\\Bar:"));
        // Should not have class attribute when using class name as ID
        assertFalse(result.contains("class: Foo\\Bar"));
    }

    public void testGenerateYamlWithClassNameAsIdFalse() {
        // generate_symfony_service_definition('Foo\\Bar', 'yaml', false)
        ServiceDefinitionGenerator generator = new ServiceDefinitionGenerator(getProject());
        String result = generator.generate("Foo\\Bar", ServiceDefinitionGenerator.OutputType.YAML, false);

        assertNotNull(result);
        // Should use snake_case service ID
        assertTrue(result.contains("foo.bar:"));
        // Should have class attribute
        assertTrue(result.contains("class: Foo\\Bar"));
    }

    public void testGenerateXmlDefault() {
        // generate_symfony_service_definition('Foo\\Bar', 'xml')
        ServiceDefinitionGenerator generator = new ServiceDefinitionGenerator(getProject());
        String result = generator.generate("Foo\\Bar", ServiceDefinitionGenerator.OutputType.XML);

        assertNotNull(result);
        assertTrue(result.contains("<service id=\"Foo\\Bar\""));
    }

    public void testGenerateXmlWithClassNameAsIdTrue() {
        // generate_symfony_service_definition('Foo\\Bar', 'xml', true)
        ServiceDefinitionGenerator generator = new ServiceDefinitionGenerator(getProject());
        String result = generator.generate("Foo\\Bar", ServiceDefinitionGenerator.OutputType.XML, true);

        assertNotNull(result);
        assertTrue(result.contains("<service id=\"Foo\\Bar\""));
        // Should not have class attribute when using class name as ID
        assertFalse(result.contains("class="));
    }

    public void testGenerateXmlWithClassNameAsIdFalse() {
        // generate_symfony_service_definition('Foo\\Bar', 'xml', false)
        ServiceDefinitionGenerator generator = new ServiceDefinitionGenerator(getProject());
        String result = generator.generate("Foo\\Bar", ServiceDefinitionGenerator.OutputType.XML, false);

        assertNotNull(result);
        // Should use snake_case service ID
        assertTrue(result.contains("id=\"foo.bar\""));
        // Should have class attribute
        assertTrue(result.contains("class=\"Foo\\Bar\""));
    }

    public void testGenerateReturnsNullForNonExistentClass() {
        // generate_symfony_service_definition('NonExistent\\Class')
        // MCP tool should throw IllegalArgumentException with "Class not found"
        ServiceDefinitionGenerator generator = new ServiceDefinitionGenerator(getProject());
        String result = generator.generate("NonExistent\\Class", ServiceDefinitionGenerator.OutputType.YAML, true);

        assertNull(result);
    }

    public void testGenerateWithLeadingBackslash() {
        // generate_symfony_service_definition('\\Foo\\Bar')
        // Should handle leading backslash gracefully
        ServiceDefinitionGenerator generator = new ServiceDefinitionGenerator(getProject());
        String result = generator.generate("\\Foo\\Bar", ServiceDefinitionGenerator.OutputType.YAML, true);

        assertNotNull(result);
        assertTrue(result.contains("Foo\\Bar:"));
    }
}

File: src/test/kotlin/fr/adrienbrault/idea/symfony2plugin/tests/mcp/GenerateSymfonyServiceDefinitionTest.kt
package fr.adrienbrault.idea.symfony2plugin.tests.mcp

import fr.adrienbrault.idea.symfony2plugin.action.ui.ServiceBuilder
import fr.adrienbrault.idea.symfony2plugin.mcp.service.ServiceDefinitionGenerator
import fr.adrienbrault.idea.symfony2plugin.mcp.toolset.ServiceDefinitionMcpToolset
import fr.adrienbrault.idea.symfony2plugin.tests.SymfonyLightCodeInsightFixtureTestCase

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class GenerateSymfonyServiceDefinitionTest : SymfonyLightCodeInsightFixtureTestCase() {
    private val serviceDefinitionGenerator = ServiceDefinitionGenerator()
    private val serviceDefinitionMcpToolset = ServiceDefinitionMcpToolset()

    @Throws(Exception::class)
    override fun setUp() {
        super.setUp()
        myFixture.configureFromExistingVirtualFile(myFixture.copyFileToProject("classes.php"))
        myFixture.addFileToProject(
            "app/config/services.xml",
            """
            <container xmlns="http://symfony.com/schema/dic/services">
                <services>
                    <service id="foo.bar" class="Foo\Bar"/>
                </services>
            </container>
            """.trimIndent()
        )
    }

    override fun getTestDataPath(): String {
        return "src/test/java/fr/adrienbrault/idea/symfony2plugin/tests/action/ui/fixtures"
    }

    fun testGenerateYamlDefault() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.Yaml)

        assertNotNull(result)
        assertTrue(result!!.contains("Foo\\Bar:"))
    }

    fun testGenerateYamlWithClassNameAsIdTrue() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.Yaml, true)

        assertNotNull(result)
        assertTrue(result!!.contains("Foo\\Bar:"))
        assertFalse(result.contains("class: Foo\\Bar"))
    }

    fun testGenerateYamlWithClassNameAsIdFalse() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.Yaml, false)

        assertNotNull(result)
        assertTrue(result!!.contains("foo.bar:"))
        assertTrue(result.contains("class: Foo\\Bar"))
    }

    fun testGenerateXmlDefault() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.XML)

        assertNotNull(result)
        assertTrue(result!!.contains("<service id=\"Foo\\Bar\""))
    }

    fun testGenerateXmlWithClassNameAsIdTrue() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.XML, true)

        assertNotNull(result)
        assertTrue(result!!.contains("<service id=\"Foo\\Bar\""))
        assertFalse(result.contains("class="))
    }

    fun testGenerateXmlWithClassNameAsIdFalse() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.XML, false)

        assertNotNull(result)
        assertTrue(result!!.contains("id=\"foo.bar\""))
        assertTrue(result.contains("class=\"Foo\\Bar\""))
    }

    fun testGenerateFluentWithClassNameAsIdTrue() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.Fluent, true)

        assertNotNull(result)
        assertTrue(result!!.contains("\$services->set(\\Foo\\Bar::class)"))
        assertFalse(result.contains(", \\Foo\\Bar::class"))
    }

    fun testGenerateFluentWithClassNameAsIdFalse() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.Fluent, false)

        assertNotNull(result)
        assertTrue(result!!.contains("\$services->set('foo.bar', \\Foo\\Bar::class)"))
    }

    fun testGeneratePhpArrayWithClassNameAsIdTrue() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.PhpArray, true)

        assertNotNull(result)
        assertTrue(result!!.contains("\\Foo\\Bar::class => ["))
        assertFalse(result.contains("'class'"))
    }

    fun testGeneratePhpArrayWithClassNameAsIdFalse() {
        val result = serviceDefinitionGenerator.generate(project, "Foo\\Bar", ServiceBuilder.OutputType.PhpArray, false)

        assertNotNull(result)
        assertTrue(result!!.contains("'foo.bar' => ["))
        assertTrue(result.contains("'class' => \\Foo\\Bar::class"))
    }

    fun testGenerateReturnsNullForNonExistentClass() {
        val result = serviceDefinitionGenerator.generate(project, "NonExistent\\Class", ServiceBuilder.OutputType.Yaml, true)

        assertNull(result)
    }

    fun testGenerateWithLeadingBackslash() {
        val result = serviceDefinitionGenerator.generate(project, "\\Foo\\Bar", ServiceBuilder.OutputType.Yaml, true)

        assertNotNull(result)
        assertTrue(result!!.contains("Foo\\Bar:"))
    }

    fun testCreateModelUsesSharedPrefillLogicForDeclaredTypes() {
        myFixture.addFileToProject(
            "union_types.php",
            """
            <?php

            namespace Foo {
                class UnionBar
                {
                    public function __construct(Bar|string ${'$'}bar)
                    {
                    }
                }
            }
            """.trimIndent()
        )

        val result = serviceDefinitionGenerator.createModel(project, "Foo\\UnionBar")

        assertNotNull(result)
        assertEquals(1, result!!.modelParameters.size)
        assertEquals("foo.bar", result.modelParameters[0].currentService)
    }

    fun testGenerateMultipleYamlDefinitionsFromCommaSeparatedClassNames() {
        myFixture.addFileToProject(
            "multiple_classes.php",
            """
            <?php

            namespace Foo {
                class Baz
                {
                }
            }
            """.trimIndent()
        )

        val result = serviceDefinitionMcpToolset.generateDefinitions(
            project,
            "Foo\\Bar, Foo\\Baz",
            ServiceBuilder.OutputType.Yaml,
            true
        )

        assertTrue(result.contains("Foo\\Bar:"))
        assertTrue(result.contains("Foo\\Baz:"))
        assertTrue(result.contains("\n---\n"))
    }

    fun testGenerateMultipleYamlDefinitionsIgnoresBlankCommaSeparatedItems() {
        myFixture.addFileToProject(
            "multiple_classes_blank_items.php",
            """
            <?php

            namespace Foo {
                class BazBlank
                {
                }
            }
            """.trimIndent()
        )

        val result = serviceDefinitionMcpToolset.generateDefinitions(
            project,
            " , Foo\\BazBlank , ",
            ServiceBuilder.OutputType.Yaml,
            true
        )

        assertEquals("Foo\\BazBlank: ~", result)
    }

    fun testGenerateMultipleYamlDefinitionsKeepsGoingWhenOneClassIsMissing() {
        val result = serviceDefinitionMcpToolset.generateDefinitions(
            project,
            "Foo\\Bar, Missing\\Class",
            ServiceBuilder.OutputType.Yaml,
            true
        )

        assertTrue(result.contains("Foo\\Bar:"))
        assertTrue(result.contains("Error: Class not found: Missing\\Class"))
        assertTrue(result.contains("\n---\n"))
    }
}
